File: src/core/test-utils.ts
import { assertSpyCalls, type Spy, spy } from "@std/testing/mock";
import type { LogFn, Logger } from "./logger.ts";
import type { Displays } from "../displays/mod.ts";
import type { Commands } from "../commands/mod.ts";
import { assertArrayIncludes, assertEquals } from "@std/assert";

export type SpyNested<
  T,
  // deno-lint-ignore no-explicit-any
  E extends (...args: any) => any = (...args: any) => any,
> = {
  [K in keyof T]: T[K] extends E ? Spy<T[K], Parameters<T[K]>, ReturnType<T[K]>>
    : T[K] extends object ? SpyNested<T[K], E>
    : never;
};

export type SpyNestedCallNumber<
  T,
  // deno-lint-ignore no-explicit-any
  E extends (...args: any) => any = (...args: any) => any,
> = Partial<
  {
    [K in keyof T]: T[K] extends E ? number
      : T[K] extends object ? SpyNestedCallNumber<T[K], E>
      : never;
  }
>;

export type SpyLogger = SpyNested<Logger, LogFn>;

export const makeSpyLogger = (): SpyLogger => ({
  log: {
    debug: spy(),
    info: spy(),
    warn: spy(),
    error: spy(),
    log: spy(),
  },
  logNoNl: {
    debug: spy(),
    info: spy(),
    warn: spy(),
    error: spy(),
    log: spy(),
  },
});

type SpyLoggerCallNumber = SpyNestedCallNumber<Logger, LogFn>;

export const assertSpyLoggerCalls = (
  logger: SpyLogger,
  callNumber?: SpyLoggerCallNumber,
) => {
  assertSpyCalls(logger.log.debug, callNumber?.log?.debug ?? 0);
  assertSpyCalls(logger.log.error, callNumber?.log?.error ?? 0);
  assertSpyCalls(logger.log.info, callNumber?.log?.info ?? 0);
  assertSpyCalls(logger.log.log, callNumber?.log?.log ?? 0);
  assertSpyCalls(logger.log.warn, callNumber?.log?.warn ?? 0);
  assertSpyCalls(logger.logNoNl.debug, callNumber?.logNoNl?.debug ?? 0);
  assertSpyCalls(logger.logNoNl.error, callNumber?.logNoNl?.error ?? 0);
  assertSpyCalls(logger.logNoNl.info, callNumber?.logNoNl?.info ?? 0);
  assertSpyCalls(logger.logNoNl.log, callNumber?.logNoNl?.log ?? 0);
  assertSpyCalls(logger.logNoNl.warn, callNumber?.logNoNl?.warn ?? 0);
};

export type SpyDisplays = SpyNested<Displays>;

export const makeSpyDisplays = (): SpyDisplays => ({
  environment: spy(),
  help: spy(),
  version: spy(),
});

type SpyDisplaysCallNumber = SpyNestedCallNumber<Displays>;

export const assertSpyDisplaysCalls = (
  displays: SpyDisplays,
  callNumber?: SpyDisplaysCallNumber,
) => {
  assertSpyCalls(displays.environment, callNumber?.environment ?? 0);
  assertSpyCalls(displays.help, callNumber?.help ?? 0);
  assertSpyCalls(displays.version, callNumber?.version ?? 0);
};

export type SpyCommands = SpyNested<Commands>;

export const makeSpyCommands = (): SpyCommands => ({
  adopt: spy(),
  apply: spy(),
  diff: spy(),
});

type SpyCommandsCallNumber = SpyNestedCallNumber<Commands>;

export const assertSpyCommandsCalls = (
  commands: SpyCommands,
  callNumber?: SpyCommandsCallNumber,
) => {
  assertSpyCalls(commands.adopt, callNumber?.adopt ?? 0);
  assertSpyCalls(commands.apply, callNumber?.apply ?? 0);
  assertSpyCalls(commands.diff, callNumber?.diff ?? 0);
};

export const assertArrayIncludesRemoveMatch = <T>(
  actual: T[],
  expected: T[],
) => {
  assertEquals(
    actual.length,
    expected.length,
    `Arrays are not the same size; actual: ${actual.length}, expected: ${expected.length}`,
  );

  for (let i = expected.length - 1; i >= 0; i -= 1) {
    assertArrayIncludes(actual, [expected[i]]);
    actual.splice(actual.indexOf(expected[i]), 1);
    expected.splice(i, 1);
  }
};

File: src/core/test-utils.go
package core

import (
	"github.com/onsi/gomega"
)

type SpyCallNoRt struct {
	Args []any
}

type SpyCalls struct {
	Debug   []SpyCallNoRt
	Debugnl []SpyCallNoRt
	Error   []SpyCallNoRt
	Errornl []SpyCallNoRt
	Info    []SpyCallNoRt
	Infonl  []SpyCallNoRt
	Log     []SpyCallNoRt
	Lognl   []SpyCallNoRt
	Warn    []SpyCallNoRt
	Warnnl  []SpyCallNoRt
}

type SpyLoggerCallNumber struct {
	Debug   int
	Debugnl int
	Error   int
	Errornl int
	Info    int
	Infonl  int
	Log     int
	Lognl   int
	Warn    int
	Warnnl  int
}

type SpyLogger struct {
	ILogger

	Calls SpyCalls
}

func (sl *SpyLogger) Debug(msg string, args ...any) {
	sl.Calls.Debug = append(sl.Calls.Debug, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Error(msg string, args ...any) {
	sl.Calls.Error = append(sl.Calls.Error, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Info(msg string, args ...any) {
	sl.Calls.Info = append(sl.Calls.Info, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Log(msg string, args ...any) {
	sl.Calls.Log = append(sl.Calls.Log, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Warn(msg string, args ...any) {
	sl.Calls.Warn = append(sl.Calls.Warn, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Debugnl(msg string, args ...any) {
	sl.Calls.Debugnl = append(sl.Calls.Debugnl, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Errornl(msg string, args ...any) {
	sl.Calls.Errornl = append(sl.Calls.Errornl, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Infonl(msg string, args ...any) {
	sl.Calls.Infonl = append(sl.Calls.Infonl, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Lognl(msg string, args ...any) {
	sl.Calls.Lognl = append(sl.Calls.Lognl, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func (sl *SpyLogger) Warnnl(msg string, args ...any) {
	sl.Calls.Warnnl = append(sl.Calls.Warnnl, SpyCallNoRt{Args: append([]any{msg}, args...)})
}

func MakeSpyLogger() *SpyLogger {
	return &SpyLogger{}
}

func AssertSpyLoggerCalls(logger SpyLogger, callNumber *SpyLoggerCallNumber) {
	var callNumberVal SpyLoggerCallNumber

	if callNumber != nil {
		callNumberVal = *callNumber
	} else {
		callNumberVal = SpyLoggerCallNumber{}
	}

	gomega.Expect(logger.Calls.Debug).To(gomega.HaveLen(callNumberVal.Debug))
	gomega.Expect(logger.Calls.Error).To(gomega.HaveLen(callNumberVal.Error))
	gomega.Expect(logger.Calls.Info).To(gomega.HaveLen(callNumberVal.Info))
	gomega.Expect(logger.Calls.Log).To(gomega.HaveLen(callNumberVal.Log))
	gomega.Expect(logger.Calls.Warn).To(gomega.HaveLen(callNumberVal.Warn))

	gomega.Expect(logger.Calls.Debugnl).To(gomega.HaveLen(callNumberVal.Debugnl))
	gomega.Expect(logger.Calls.Errornl).To(gomega.HaveLen(callNumberVal.Errornl))
	gomega.Expect(logger.Calls.Infonl).To(gomega.HaveLen(callNumberVal.Infonl))
	gomega.Expect(logger.Calls.Lognl).To(gomega.HaveLen(callNumberVal.Lognl))
	gomega.Expect(logger.Calls.Warnnl).To(gomega.HaveLen(callNumberVal.Warnnl))
}
